### DISCUSSION TITLE
Could cockpit be useful on nixos?

### DISCUSSION THREAD

**Discussion by pbsds** [cockpit-project/cockpit#17558] [General]
Nixos is a distribution famed for defining the whole system state with its system derivation / config file.
Things like which systemd services, containers and VMs should be running, which mountpoints and users to have and so on.
If i were to setup cockpit on nixos, would it provide value still?

**Comment by martinpitt** [comment]
I don't have much first-hand experience with nix (just with CI for umockdev). I suppose it would still be useful as a "read-only" monitoring tool, so that you can debug why your machine is acting up (such as, failed services, running out of RAM/disk, biggest memory consumers, reading journal, etc.) You may not necessarily be able to change things due to a read-only OS. But I'd still expect e.g. the systemd D-Bus service to work, so that you can stop and start units and such?

But at this point it's just a very high-level gut feeling. So far I've mostly seen nix as a container base, but apparently you can even [run a desktop with it](https://nixos.wiki/wiki/NixOS_as_a_desktop). I think someone needs to try and build/run it and see what goes wrong :grin:

**Comment by astratagem** [comment]
> I think someone needs to try and build/run it and see what goes wrong 😁

https://github.com/NixOS/nixpkgs/issues/38161

I haven’t used Cockpit at all, but I’d be interested to see it working on NixOS too. Based on the discussion linked above, it looks like some other NixOS users have made progress.

**Comment by astratagem** [comment]
There is now a NixOS module for Cockpit:

https://github.com/NixOS/nixpkgs/pull/209192

**Comment by dgiridharan** [comment]
Amazing! Thanks for providing the NixOS module for Cockpit!

I wanted to access Cockpit from another machine (and not just localhost) within the network. As such I've the following configuration:

```
services.cockpit = {
  enable = true;
  port = 9090;
  openFirewall = true;
  settings = {
    WebService = {
      AllowUnencrypted = true;
    };
  };
};
```
Everything works well. That said, when I try accessing Cockpit from my laptop on the same network, I can't enable admin controls. It throws the following error: 
![image](https://github.com/user-attachments/assets/413940e5-93a2-4d72-a484-80f1a8c87b01)

I'm really new to Nix so sorry if this is a basic problem!

I suspect this might be due to different paths on NixOS. Does anyone know if it’s possible to use [nix-ld](https://github.com/nix-community/nix-ld) to resolve this issue?

Would love to hear if anyone has managed to fix this. Thanks in advance!

**Reply by jelly** [reply]
Normally this would be /usr/libexec or /usr/lib/cockpit/. So it sounds like this is a packaging bug in NixOS. 

Seeing:

```
        ExecStart = "${cfg.package}/libexec/cockpit-ws --for-tls-proxy --port=0";
```

Used, does `/$whatever/libexec/cockpit-ws` exist in NixOS

**Reply by pbsds** [reply]
`/nonexistent` is likely the `$HOME` variable set by systemd's `ProtectHome` flag.

**Reply by lucasew** [reply]
> Normally this would be /usr/libexec or /usr/lib/cockpit/. So it sounds like this is a packaging bug in NixOS.
> 
> Seeing:
> 
> ```
>         ExecStart = "${cfg.package}/libexec/cockpit-ws --for-tls-proxy --port=0";
> ```
> 
> Used, does `/$whatever/libexec/cockpit-ws` exist in NixOS

This $whatever would be /run/current-system/sw AFAIK
